docs/WASI-design-principles.md

```diff
@@ -0,0 +1,93 @@
+
+Compatibility with existing host environments and applications is important, and we have put a lot of work into WASI libc to provide POSIX compatibility and support existing applications. There's a lot more work that can be done, and a lot more we can do to improve compatibility and user convenience. We're continuing to make progress -- and users are encouraged to [file bugs](https://github.com/WebAssembly/WASI/issues) when they find things that don't work or are awkward. This approach supports existing applications, while also supporting applications and libraries willing to opt in to enable stronger and more fine-grained security properties than are possible in regular POSIX.
+
+We recognize that this approach has tradeoffs. It often does take more work to design and implement the compatibility layers needed to support existing applications than if we just always made WASI always expose POSIX-style APIs directly. And, even when we do have compatibility mechanisms, they aren't always the most locally optimal ones. The compatibility layer overhead is usually quite modest, but it is present.
+
+However, shared-nothing linking makes it easier to use libraries in more circumstances, because you don't have to have the trust implied by a shared filesystem view, or the complexity of configuring filesystem rules for each library. With a better story for libraries and tools to work together in cooperation with the sandbox, we can build a stronger ecosystem which makes up for the downsides in the long run.
+
+## Relationship to the Web
+
+It is possible to run WASI code on the Web with the help of polyfills, however WASI isn't limited to APIs which run well or are easy or efficient to polyfill on the Web.
+
+That said, where other considerations don't interfere, WASI should use existing Web standards and work with Web standardization efforts rather than gratuitously inventing its own versions of them and/or duplicating efforts.
+
+When using Web standards, WASI APIs should be careful to avoid depending on JavaScript in the engine in APIs where it isn't essential.
+
+## Use WebAssembly standards and proposals
+
+WASI should align with and build on WebAssembly standards and proposals where applicable.
+
+For example, WASI seeks to align with and build on [interface types](https://github.com/WebAssembly/interface-types), [multiple return values](https://github.com/WebAssembly/multi-value/), [reference types](https://github.com/WebAssembly/reference-types), [type imports](http://github.com/WebAssembly/type-imports-proposal), and more. As of this writing, some of these are early-stage proposals, so we're not actually depending on them yet, however we are carefully aligning with them so that we'll be ready when they are.
```

Comment:
noticed a broken link, should be https://github.com/WebAssembly/proposal-type-imports

Reply:
Fixed, thanks!